### PR TITLE
cache: add error handling to informers

### LINKED ISSUE #155: Informers do not surface API server request failures to callers
When I call [`cache.NewInformer`](https://github.com/kubernetes/client-go/blob/65ac4e638a79d49b7dbd5e1cef4e7785af7cbd39/tools/cache/controller.go#L235-L240) to create an Informer and then call [`cache.Controller.Run`](https://github.com/kubernetes/client-go/blob/65ac4e638a79d49b7dbd5e1cef4e7785af7cbd39/tools/cache/controller.go#L235-L240) on the returned value, the controller periodically contacts the API server to list the objects of the given type. At some point the API server rejects these requests, causing the controller to log messages like the following:
> Failed to list *v1beta1.Ingress: the server has asked for the client to provide credentials (get ingresses.extensions)\n","stream":"stderr","time":"2017-03-21T14:14:37.397681364Z"}
>
> Failed to list *v1.Service: the server has asked for the client to provide credentials (get services)\n","stream":"stderr","time":"2017-03-21T14:14:37.398311081Z"}
>
> Failed to list *v1.Endpoints: the server has asked for the client to provide credentials (get endpoints)\n","stream":"stderr","time":"2017-03-21T14:14:37.399217817Z"}

These failures repeat periodically, swelling the log file, and it may take many days before we notice that our controller is ostensibly still running, but only inertly; it can't get its job done without talking to the API server. Sometimes the server changes its mind and starts fulfilling the requests again, but these failure periods can persist for days.

_A caller of `cache.Controller.Run` should have some way of detecting that these failures are occurring in order to declare the process unhealthy._ Retrying automatically to smooth over intermittent network trouble is a nice feature, but with neither the ability to control it nor detect its ongoing failure makes it dangerous.

I would be happy with either of the following two improvements:
- **Accept a callback (perhaps via a new sibling method for `Controller.Run`) that tells a caller when these request failures arrive.**  
  It could also accept a caller-provided channel, and push errors into the channel when they arise, dropping errors that can't be delivered synchronously.
- **Provide a way to integrate a controller into a "healthz" handler.**  
  That leaves the health criteria opaque to callers—and probably begs for some way to configure the thresholds—but still allows a calling process to indicate that it's in dire shape.

We discussed this gap [in the "kubernetes-dev" channel in the "Kubernetes" Slack team](https://kubernetes.slack.com/archives/C09R23FHP/p1490107292534753).

### PR DESCRIPTION
**Pull request by nicks**
When creating an informer, this adds a way to add custom error handling
or backoff logic, so that Kubernetes tooling can properly surface
the errors to the terminal.

Fixes https://github.com/kubernetes/client-go/issues/155

/kind feature

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -327,7 +342,7 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 			case err == io.ErrUnexpectedEOF:
 				klog.V(1).Infof("%s: Watch for %v closed with unexpected EOF: %v", r.name, r.expectedTypeName, err)
 			default:
-				utilruntime.HandleError(fmt.Errorf("%s: Failed to watch %v: %v", r.name, r.expectedTypeName, err))
+				errToReturn = fmt.Errorf("Failed to watch %v: %v", r.expectedTypeName, err)
```
Comment (discordianfish) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: Why not return the error right away?

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 2]
```diff
@@ -318,6 +332,7 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 
 		w, err := r.listerWatcher.Watch(options)
 		if err != nil {
```
Comment (discordianfish) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: What about removing all the specific error handling here to a DefaultDropWatchHandler that is registered by default?
Because I'd expect people would like to handle EOF and ConnectionRefused error (if only for monitoring purposes).

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 1 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: i wanted to preserve the current behavior where we don't return on connection refused (i.e., line 351)

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 2 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: The most important thing for us is to have some mechanism to bubble up to user "hey, your auth tokens expired, and we can't talk to kubernetes anymore".

I'm deeply worried that this PR is going to turn into a quagmire where we can't add any error-handling mechanism at all until we've had a long discussion on which errors bubble up and which do not. 

I think that's a good discussion to have. But I don't feel like I'm well-equipped to facilitate that discussion, or to weigh competing needs (e.g., people who want the informer to retry EOF internally vs those who do not)

Comment (discordianfish) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: Understood, but I'm just suggesting to keep the existing error handling but move it to a DefaultDropWatchHandler. That way a user can customize it without changing the existing default behaviour.

Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: hmmm...i spent some time playing around with this, and couldn't come up with a solution that fit well. This error handling depends on unexported fields and functions of the package (r.expectedTypeName, isExpiredError), and wasn't sure how much we should really be exposing.

Did you have a particular API in mind?

Comment (discordianfish) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: @nicks A DefaultDropWatchHandler could use these fields, right? Something like this:
```
func DefaultDropWatchHandler(err error, doneCh <-chan struct{}) {
	switch {
			case isExpiredError(err):
				r.setIsLastSyncResourceVersionExpired(true)
				klog.V(4).Infof("%s: watch of %v closed with: %v", r.name, r.expectedTypeName, err)
			case err == io.EOF:
				// watch closed normally
			case err == io.ErrUnexpectedEOF:
				klog.V(1).Infof("%s: Watch for %v closed with unexpected EOF: %v", r.name, r.expectedTypeName, err)
			default:
				utilruntime.HandleError(fmt.Errorf("%s: Failed to watch %v: %v", r.name, r.expectedTypeName, err))
	}
	if !utilnet.IsConnectionRefused(err) {
		doneCh <- <- struct{}{}
	}
}
```

Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: ok, i pushed a new branch to demonstrate what that would look like, what do you think?

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 3]
```diff
@@ -95,6 +95,30 @@ type Reflector struct {
...
 	WatchListPageSize int64
+	// Called whenever the ListAndWatch drops a connection, whether or not
+	// the connection had an error. So that clients can handle errors
+	// and implement backoff.
+	dropWatchHandler DropWatchHandler
+}
+
+type DropWatchHandler func(err error, doneCh <-chan struct{})
+
+func createDefaultDropWatchHandler(r *Reflector) DropWatchHandler {
```
Comment (discordianfish) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: I'd change the `DropWatchHandler` type to `func(r *Reflector, err error, doneCh <- chan struct{})` and then define `func DefaultDropWatchHandler(...)`. Seems more idiomatic to me and having access to Reflector in custom handlers seems useful anyway.

Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: done!

**Review by discordianfish — APPROVED** [review]
LGTM (fwiw)

**Comment by discordianfish** [comment]
@deads2k bump

File: staging/src/k8s.io/client-go/tools/cache/controller.go [hunk 4]
```diff
@@ -69,6 +69,11 @@ type Config struct {
 	//       question to this interface as a parameter.  This is probably moot
 	//       now that this functionality appears at a higher level.
 	RetryOnError bool
+
+	// Called whenever the ListAndWatch drops a connection, whether or not
+	// the connection had an error. So that clients can handle errors
+	// and implement backoff.
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/cache/controller.go]: Given that the machinery retries, what does "implement backoff" mean here?

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 5]
```diff
@@ -95,6 +95,28 @@ type Reflector struct {
 	// etcd, which is significantly less efficient and may lead to serious performance and
 	// scalability problems.
 	WatchListPageSize int64
+	// Called whenever the ListAndWatch drops a connection, whether or not
+	// the connection had an error. So that clients can handle errors
+	// and implement backoff.
+	dropWatchHandler DropWatchHandler
+}
+
+type DropWatchHandler func(r *Reflector, err error, doneCh <-chan struct{})
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: Why is "doneCh" passed here?

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: The retry strategy is currently very naive, and isn't great in cases that require user intervention, see discussion here:
https://github.com/kubernetes/client-go/issues/155
and here:
https://github.com/windmilleng/tilt/issues/2702

In particular, MacOS users were complaining that the exec auth provider was bringing up the macos keychain dialog 1x per second, resulting in thousands of dialogs if they stepped away for a bit.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 5 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: this is kind of related to the above question: we want to make the backoff longer, or pause it entirely until manual user intervention.

But we also want to respect API guarantees around the case where the caller closes the done channel. does that make sense?

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: Triggering the exec auth callout should probably be directly rate limited or controlled. I doubt this is the only way to call it inappropriately often. Also, a call to that should be blocking (otherwise you lack something necessary to make a network call at all), so I don't understand how this retry loop could trigger it so many times.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 5 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: Maybe--it definitely needs to be documented here what a good DropWatchHandler should do.

Did you consider letting people specify the retry frequency instead? We have a number of "backoff" mechanisms that are pluggable elsewhere. If all you want to do is control the frequency / retry behavior, that seems like a more natural intervention point.

Personally, I wouldn't have expected a handler like this to be blocking, I'd expect it to just be a notification.

Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: e.g., I think it'd be very reasonable for a "backoff" object to be passed to the reflector constructor.

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: It sounds like you're both saying:
1) This fix shouldn't be necessary, and
2) Even if it was necessary, this wouldn't be the right way to fix it.

This makes me think we don't agree on what the diagnosis of the bug is (which is ok! I'm not  convinced I've diagnosed it correctly either!). 

But it seems hard to discuss whether the solution is correct if we don't agree on the diagnosis. What would help us reach agreement on the diagnosis? Are there additional tests that would help?

And then there's the meta question - is it worth spending more time on this PR? If the API is architecturally at odds with the direction this package is going, then I'm happy to cut my sunk costs now, no hard feelings.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 5 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: The problem I'm trying to solve isn't "Alice thinks the backoff should be 3 seconds, and Bob thinks the backoff should be 5 seconds, so let's make it an option!"

The problem is that this code isn't even the right layer of the API to handle errors. Is the end user looking at a terminal? Are they looking at a web dashboard? Or something other UI entirely? This code doesn't know where to send the error message or what additional context it needs.

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (discordianfish) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: I feel like this discussion should be out of scope of this PR. It doesn't change anything in that regard, it merely makes it possible to handle the error. Could that be addressed in a separate PR/issue? I think as it is, this PR is a great improvement.

Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: Apologies for lack of clarity. I am saying neither of those things.

a) I think there should be a separate rate limiting fix for the auth token getting behavior. Something is weird there that isn't explained by the code here AFAICT. E.g. if the problem is failing to wait for the human to actually click "yes" or "no" then changing this code at best reduces the frequency of that condition.
b) I agree 100% that it makes sense for the retry behavior to be controllable. We have a mostly standardish way of doing that (user constructs and passes a backoff type object).
c) I also agree 100% that it should be discoverable when this thing is not actually connected, and the reason (e.g. most recent error & count).

Where I get off the bus is that I don't think it's good to combine b and c.
* Client-go API changes affect a lot of people and it is worth our time to get them right
* Combining the error reporting and backoff behavior in this way is not consistent with what we do elsewhere in the client
* This code is very tricky; an incorrect delay in this function will make the client much less efficient (the RV is time-sensitive; if re-used rapidly, a watch can be resumed, otherwise a very expensive list has to be done)
* It's very plausible that when the connection is lost, we actually want to mark the cache/informer as "not synced" to prevent controllers from acting with this information

I don't expect a change to perform all of those functions simultaneously, but it needs to not change the API in a way that makes those things harder. Hopefully that makes sense.

In terms of concrete suggestions for this change-- I'm mostly OK with this except for the part where people use this function to do rate limiting. I can imagine either a version of this change where the function is not set up to be delayed (no channel, warning in comment not to do slow processing, etc), or a version where errors are sent through a channel. I think it will be important for the consumer of these channels to also know when we are reconnected, so that they can restart their processing. We can wait for users before adding that, but again the api needs to not rule it out (e.g. OnDisconnection/OnReconnection function pair makes sense).

You can appeal to e.g. @deads2k or @liggitt if you think I'm being too pedantic here, but given the amount of time we've spent making bug fixes to this code, I think I'm being very reasonably cautious.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 5 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: I 100% agree with that problem and this is close to being a fix--if you:
a) remove the channel
b) either add a comment explaining this is for error visibility and that one should offload any expensive processing OR launch this in a goroutine (I prefer the former)
c) Maybe change the name, ConnectionErrorHandler or WatchErrorHandler?

Then I can sign off on this.

(I think the other items from my last comment also deserve separate fixes)

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: Thank you for the long explanation! It's very helpful. 

One thing I have trouble reconciling - you said "Triggering the exec auth callout should probably be directly rate limited or controlled", but you also said "an incorrect delay in this function will make the client much less efficient". Do you think those are directly opposed constraints? 

I'm asking because I want to make sure that the comments I add address the trade-offs correctly -- would adding more rate limiting to the auth / request layer have bad outcomes for the watch code?

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 5 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: sounds good, I'll try to make this change later this afternoon

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: > "Triggering the exec auth callout should probably be directly rate limited or controlled", but you also said "an incorrect delay in this function will make the client much less efficient"

The "this" in the latter sentence is talking about the reflector retry loop, not the one needed for the auth token refresh. The need for rate limiting token refreshes is troubling because nothing in our stack is prepared for that to be anything other than instantaneous. We should probably actually start with wrapping that call with https://godoc.org/golang.org/x/sync/singleflight to prevent duplicate calls. That might actually be enough to solve that problem. @mikedanese might have another idea?

Comment (mikedanese) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: Can you point me to the MacOS exec auth plugin if it's opensource? Does every call require user interaction? How long do these processes run? How long are the credentials valid for? Clearly, 1000 pop ups is a terrible experience but I would like to have a better idea of what the most desirable behavior would be.

I imagine that this is one of many sites that does not handle this type of blocking interruption well. If we want to support credential plugins that require user interaction, we should do our best to solve these problems at (or closer to) the credential source to limit the impact and complexity we push down to the rest of the client.

Singleflighting and evaluating timeout behavior of the exec credential source sound very beneficial. Most of the mileage we've put on that extension point uses non-interactive login flows so I would be surprised if there weren't rough spots. Also :+1:, thanks for helping prove this use case out.

Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: Unfortunately, I don't know : (

The end-user complaining about this wasn't able to repro it reliably. They would take screenshots and logs and relay it to their infrastructure team at their company. Their infrastructure team would complain about it to me, who does not work for that company. It would probably take a while just to explain the question you're asking.

My personal experience with MacOS syscalls is that it can be often non-trivial to figure out when a syscall is going to pop up a dialog box or print something to the terminal.

Sorry, I appreciate that's a frustrating answer.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 5 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: @lavalamp all comments addressed! PTAL?

I'll follow up later with a PR for the backoff parameter

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 6]
```diff
@@ -275,7 +297,7 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 		case <-listCh:
 		}
 		if err != nil {
-			return fmt.Errorf("%s: Failed to list %v: %v", r.name, r.expectedTypeName, err)
+			return fmt.Errorf("Failed to list %v: %v", r.expectedTypeName, err)
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: lower case error strings here and bellow.

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 7]
```diff
@@ -174,8 +195,9 @@ var internalPackages = []string{"client-go/tools/cache/"}
 func (r *Reflector) Run(stopCh <-chan struct{}) {
 	klog.V(2).Infof("Starting reflector %s (%s) from %s", r.expectedTypeName, r.resyncPeriod, r.name)
 	wait.BackoffUntil(func() {
-		if err := r.ListAndWatch(stopCh); err != nil {
-			utilruntime.HandleError(err)
+		err := r.ListAndWatch(stopCh)
+		if err != nil {
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: Why did you change the scoping of err?

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 8]
```diff
@@ -366,19 +388,6 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
...
-				// Don't set LastSyncResourceVersionExpired - LIST call with ResourceVersion=RV already
-				// has a semantic that it returns data at least as fresh as provided RV.
-				// So first try to LIST with setting RV to resource version of last observed object.
-				klog.V(4).Infof("%s: watch of %v closed with: %v", r.name, r.expectedTypeName, err)
-			case err == io.EOF:
-				// watch closed normally
-			case err == io.ErrUnexpectedEOF:
-				klog.V(1).Infof("%s: Watch for %v closed with unexpected EOF: %v", r.name, r.expectedTypeName, err)
-			default:
-				utilruntime.HandleError(fmt.Errorf("%s: Failed to watch %v: %v", r.name, r.expectedTypeName, err))
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: This is the only case that is abnormal. If the goal is to expose unhandled internal errors, why would we forward expected errors to a custom handler?

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 7 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: haha, this PR has gone thru a bunch of revisions, and this change was from an earlier one where it made more sense. changed back!

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 5 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: > b) either add a comment explaining this is for error visibility and that one should offload any expensive processing OR launch this in a goroutine (I prefer the former)

Another approach is to attempt a non-blocking write of the error to a notification chan.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 6 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: hmmmm...I didn't add this error string in this PR. Would you accept this as a separate PR? This PR has been open for a couple months and has been more controversial than I expected, which makes me nervous about widening the scope to include unrelated changes.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 8 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: Your suggestion is what I did in the original version of this PR:
https://github.com/kubernetes/kubernetes/commit/13f9ad7cf008eba732157603701118a7997cf3cc

See discussion further up in the PR. I don't have a well-informed opinion on what the right answer is.

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 9]
```diff
@@ -163,6 +163,18 @@ type SharedInformer interface {
...
+	// connection with an error. After calling this handler, the informer
+	// will backoff and retry.
+	//
+	// The default implementation looks at the error type and tries to log
+	// the error message at an appropriate level.
+	//
+	// Implementations of this handler may display the error message in other
+	// ways. Implementations should return quickly - any expensive processing
+	// should be offloaded.
+	SetWatchErrorHandler(handler WatchErrorHandler) error
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: Shared informers are usually shared :). This interface should attempt to reflect that ownership model by exposing functionality that is safe to consume at multiple sites across potentially large applications. SharedInformers differ a lot from regular informers to make sure that's the case (see cache mutation detection, and shared processor). This function stands out to me because it is:
* Exclusive: last call before starting the informer wins. Previous calls are silently ignored.
* Has potential for unknown/unbounded/hard to notice spillover effects on other users of the same shared informer.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 6 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: According to my diff, you removed the initial %s which printed a name? So I think this is a fair request. Below errors, too. We're currently in code freeze at the moment, so please expect you'll end up rebasing one or more times before this merges anyway.

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go (hunk 9 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: I don't think this is the mechanism we'll use to get controllers to e.g. stop operating during a disconnection. This is just for visibility.

I think we can state these things in the comment and that will be sufficient.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 10]
```diff
@@ -95,6 +95,26 @@ type Reflector struct {
 	// etcd, which is significantly less efficient and may lead to serious performance and
 	// scalability problems.
 	WatchListPageSize int64
+	// Called whenever the ListAndWatch drops the connection with an error.
+	watchErrorHandler WatchErrorHandler
+}
+
+type WatchErrorHandler func(r *Reflector, err error)
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: I'd repeat your disclaimer (about not doing work that will take time) in a comment here, as well.

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: Even (especially?) in the face of not knowing how exactly it works, I think it makes sense to singleflight those calls.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 6 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: (And I see this PR had been open a long time before I got to it. Sorry about that.)

File: staging/src/k8s.io/client-go/tools/cache/controller.go (hunk 4 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: A single authenticator *should* already be singleflighted. At least it's supposed to be. I want to check if multiple are getting instantiated in some of the core binaries or if there's something going on.

Comment (mikedanese) [reply staging/src/k8s.io/client-go/tools/cache/controller.go]: I filed https://github.com/kubernetes/kubernetes/issues/89114

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 10 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: done

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 6 above)
Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: makes sense to me, done!

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go (hunk 9 above)
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: Sorry for not being clear, I think we should add to the comment here:

* There's only one handler, so if you call this multiple times, last one wins; calling after the informer has been started returns an error.
* The handler is intended for visibility, not to e.g. pause the consumers.

(these points don't need to go in the place where you copied this comment, just here--since they're unique to shared informers, or at least the first one is.)

Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: sorry - github aggressively hides conversations marked 'resolved' so i missed that this was still outstanding! I like your suggestions, added!

**Comment by lavalamp** [comment]
/lgtm
/approve

Thanks!

**Comment by lavalamp** [comment]
`verify` is almost never flaky, btw.

**Comment by nicks** [comment]
/test pull-kubernetes-verify

@lavalamp I 100% believe you that verify is not flaky, but is there a guide anywhere on how to read its output?

The logs both seem to be saying "all tests passed" and the job exited with failure.
 https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/87329/pull-kubernetes-verify/1239582059362521088

**Comment by nicks** [comment]
/retest

oh, nm, it passed!!

**Comment by lavalamp** [comment]
I think it actually is flaky right now. Oops! The one time I don't check...

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 10 above)
Comment (tedyu) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:112]: Should the caller put a cap on the amount of time the handler executes ?
Just in case some implementation doesn't follow the guide above.

Comment (nicks) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:112]: @tedyu can you point to an established pattern for doing that safely? were you thinking a done channel to the handler, or something else?

Comment (tedyu) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:112]: Yeah, similar to done channel.
You can create goroutine for running the handler.
In the caller, use a timer to gauge the amount of time handler runs.

Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:112]: I think that is too much complexity for this. If people abuse this, we can fix it. Until then, just documentation, please.

**Comment by Levovar** [comment]
@nicks not trying to necro the PR but considering almost all but two members of cache.Reflector are unexported:
https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L49
not sure how useful it is to reguire it in WatchErrorHandler's signature
I was happy very with hit new API and planning to use it in my Informers, but I feel this input parameter will be mostly unused due to this
what do you think about it?
